File: src/main/scala/myUtil/NTT.scala
package myUtil

import scala.collection.mutable.ArrayBuffer
import scala.math
import scala.util.Random

class NTT {
  val mod = 65537
  val g = 3
  val polyLen = 4

  def bigNumModPow(a: Int, b: Int, mod: Int): Int = {
    if (b == 0) {
      return 1
    }
    var res:BigInt = 1
    for (i <- 0 until b) {
      res *= a
      res %= mod
    }
    res.toInt
  }

  def twiddleMatrix(polyLen: Int, root: Int, mod: Int):Array[Array[Int]] = {
    val mat = Array.fill(polyLen)(Array.fill(polyLen)(0))
    for (i <- 0 until polyLen){
      val rootI = bigNumModPow(root, i, mod)
      for (j <- 0 until polyLen){
        val rootIJ = bigNumModPow(rootI,j,mod)
//        println(s"${i} ${j} ${rootI} ${rootIJ}")
        mat(i)(j) = rootIJ
      }
    }
    mat
  }

  def getRoot(n:Int = polyLen):Int={
    bigNumModPow(g, (mod - 1)/n, mod)
  }

  def getRandomInput(): Array[Int] = {
//    Random.setSeed(2)
    Array.fill(polyLen)(Random.nextInt(mod))
  }

  def getTwiddleMatrix(): Array[Array[Int]] = {
    twiddleMatrix(polyLen, getRoot(), mod)
  }

  def multiply(matrix0: Array[Array[Int]], matrix1: Array[Array[Int]], resWidth: Int) = {
    assert(matrix1.length == matrix0.length)
    val numRows = matrix0.length
    val resultMatrix = Array.fill(numRows)(Array.fill(numRows)(BigInt(0)))
    val maxValue:BigInt = (BigInt(1) << resWidth) - 1
    for (i <- 0 until numRows) {
      for (j <- 0 until numRows) {
        var sum:BigInt = 0
        for (k <- 0 until numRows) {
          val a = BigInt(matrix0(i)(k))
          val b = BigInt(matrix1(k)(j))
          sum += a * b
        }
        resultMatrix(i)(j) = sum.min(maxValue) % mod
      }
    }
    resultMatrix
  }
}

object NTT extends App {
  val ntt = new NTT
  val root = ntt.getRoot()
  println(root)
  val inputMat = Array.fill(ntt.polyLen)(ntt.getRandomInput()).transpose
  val twiddleMat = ntt.getTwiddleMatrix()
  import PrintDump._
  val res = ntt.multiply(twiddleMat, inputMat, 64)
  printMatrix(inputMat,"inputMatrix")
  printMatrix(twiddleMat,"twiddleMatrix")
  printMatrix(res,"result")

}

File: src/main/ntt.py
import random
from sympy import ntt

# 快速幂
def big_num_mod_pow(a, b, mod):
    if b == 0:
        return 1
    res = 1
    for i in range(b):
        res *= a
        res %= mod
    return res

# 旋转因子
def vandermonde(polyLen, root, mod):
    matrix = [[] for i in range(polyLen)]
    for i in range(polyLen):
        root_i = big_num_mod_pow(root, i, mod)
        for j in range(polyLen):
            root_i_j = big_num_mod_pow(root_i, j, mod)
            matrix[i].append(root_i_j)
    return matrix


mod = 65537
g = 3


root_16 = big_num_mod_pow(g, (mod-1)//16, mod)

# inputArray_16 = [random.randint(0, mod) for i in range(16)]
# res_16 = ntt(inputArray_16, mod)

root_4 = big_num_mod_pow(g,(mod-1)//4,mod)
inputArray_4 = [56199, 14191, 12134,12035]
res_4 = ntt(inputArray_4, mod)


# print("vandermonde_16: ", vandermonde(16, root_16, mod))
# print("inputArray_16: ", inputArray_16)
# print("res_16: ", res_16)

print("vandermonde_4: ", vandermonde(4, root_4, mod))
print("inputArray_4: ", inputArray_4)
print("res_4: ", res_4)